batch/batch/server/database.py

```diff
@@ -0,0 +1,167 @@
+        assert all([k in schema for k in keys])
+
+        async with self.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                schema = ", ".join([f"`{n}` {t}" for n, t in schema.items()])
+                key_names = ", ".join([f'`{name.replace("`", "``")}`' for name in keys])
+                keys = f", PRIMARY KEY( {key_names} )" if keys else ''
+                sql = f"CREATE TABLE IF NOT EXISTS `{name}` ( {schema} {keys})"
+                await cursor.execute(sql)
+
+    def create_table_sync(self, name, schema, keys):
+        return run_synchronous(self.create_table(name, schema, keys))
+
+    async def create_temporary_table(self, name, schema, keys):
+        name = await self.temp_table_name(name)
+        t = await Table(self, name, schema, keys)
+        return t
+
+    def create_temporary_table_sync(self, name, schema, keys):
+        return run_synchronous(self.create_temporary_table(name, schema, keys))
+
+
+def make_where_statement(items):
+    template = []
+    values = []
+    for k, v in items.items():
+        if isinstance(v, list):
+            if len(v) == 0:
+                template.append("FALSE")
+            else:
+                template.append(f'`{k.replace("`", "``")}` IN %s')
+                values.append(v)
+        else:
+            template.append(f'`{k.replace("`", "``")}` = %s')
+            values.append(v)
+
+    template = " AND ".join(template)
+    return template, values
+
+
+@asyncinit
+class Table:  # pylint: disable=R0903
+    async def __init__(self, db, name, schema, keys):
+        self.name = name
+        self._db = db
+        await self._db.create_table(name, schema, keys)
+
+    async def _new_record(self, items):
+        names = ", ".join([f'`{name.replace("`", "``")}`' for name in items.keys()])
+        values_template = ", ".join(["%s" for _ in items.values()])
+        async with self._db.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                sql = f"INSERT INTO `{self.name}` ({names}) VALUES ({values_template})"
+                await cursor.execute(sql, tuple(items.values()))
+                id = cursor.lastrowid  # This returns 0 unless an autoincrement field is in the table
+        return id
+
+    async def _update_record(self, where_items, set_items):
+        async with self._db.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                if len(set_items) != 0:
+                    where_template, where_values = make_where_statement(where_items)
+                    set_template = ", ".join([f'`{k.replace("`", "``")}` = %s' for k, v in set_items.items()])
+                    set_values = set_items.values()
+                    sql = f"UPDATE `{self.name}` SET {set_template} WHERE {where_template}"
+                    await cursor.execute(sql, (*set_values, *where_values))
+
+    async def _get_records(self, where_items, select_fields=None):
```

Comment:
General question: Why do these methods have private-method semantics?

Reply:
Because I override some of them in code that's coming in a future PR to make the interface nicer. For example, 

```
class BatchJobsTable(Table):
    async def __init__(self, db, name='batch-jobs'):
        schema = {'id': 'BIGINT',
                  'job': 'BIGINT'}
        keys = []

        await super().__init__(db, name, schema, keys)

    async def new_record(self, **items):
        await self._new_record(items)

    async def get_jobs(self, id):
        result = await self._get_records({'id': id})
        return [record['job'] for record in result]

    async def delete_record(self, batch_id, job_id):
        await self._delete_records({'id': batch_id,
                                    'job': job_id})

    async def has_record(self, batch_id, job_id):
        return await self._has_record({'id': batch_id,
                                       'job': job_id})
```

Reply:
What is the advantage over `super(BatchJobsTable, self).new_record(**items)` or `Table.new_record(self, **items)`